### ISSUE TITLE
Syntax code blocks don't flow over page boundaries

### ISSUE THREAD

**Issue by wolandscat** [asciidoctor/asciidoctor-pdf#240]
If a code block in the adoc source doesn't all fit in the remaining space in the page in the PDF where it occurs, it gets pushed to the next page, leaving a large empty space. What happens with a really long code block?

**Comment by mojavelinux** [comment]
The current logic is that it does it's best to keep the block together. Once the block spans more than one page, it starts splitting the code block across pages without first advancing to a new page.

What we problem need is something like an orphans setting. A code block should try to fit up to a certain number of lines (or percentage) and then split. The logic for splitting wasn't as advanced as it is now, so we can now start to explore being less conservative (i.e., splitting earlier).

**Comment by wolandscat** [comment]
Yes - orphans would be the normal DTP approach to this. In Frame, you set # of orphan rows, normal default is 2. So the logic should probably be something like (pseudocode):

```
lines_available := available space left on page as a line count
block_lines := number of lines of current code block
if blocklines mod lines_available < ORPHAN_LINES -- modulo operator
    start a new page
end

start outputting lines normally...
```

**Comment by shankardevy** [comment]
Is there any update to this issue? I also need the source code block to start in the immediately available space, rather than introducing a page break which gives large amount of unnecessary white spaces.

**Comment by alvarolop** [comment]
I also think that this feature could improve the quality of my documents. Is there any estimated time of arrival for this functionality?

**Comment by RarrugasTetra** [comment]
I am also dependent on this issue. Hopefully this is fixed in near future.

**Comment by mojavelinux** [comment]
I've been working on it for the past few weeks. I'm very close to having it working.

**Comment by gkarabotsos** [comment]
Any updates on this?
Ty!

**Comment by mojavelinux** [comment]
I'm back to working on it. I'm going to do everything I can to get this finished and get an alpha release out in the next few weeks.

**Comment by graphitefriction** [comment]
All delimited blocks are breakable by default in the upcoming Asciidoctor PDF 2.0 release. When a delimited block doesn't fit on the current page and doesn't occupy more than one page, it will start on the current page, break, and then continue on the next page. If you ever want a specific block to not break, you'll apply the `unbreakable` option to the block.

```
[source%unbreakable,java]
----
This is a delimited source block.
----
```

This issue is related to #2004 and was resolved by PR #2020.
